File: Code/Development/NeuralNet/Experiment.m
classdef Experiment < handle
    properties
        sources
    end
    methods
        function obj = Experiment()
            fprintf('\nLoading data.\n');
            obj.sources = {};
            %obj.sources{end+1} = HousingData('Data/nashville_processed.csv', 'Nashville');
            %obj.sources{end+1} = HousingData('Data/kingcounty_processed.csv', 'KingCounty');
            %obj.sources{end+1} = HousingData('Data/redfin_processed.csv', 'GrandRapids');
            obj.sources{end+1} = HousingData('Data/art_processed.csv', 'ART');
        end
        function run(obj)
            n_sources = length(obj.sources);
            for i = 1:n_sources
                source = obj.sources{i};
                
                fprintf('\nData ----------------------------------\n');
                [dataset_train, dataset_test] = source.to_dataset().split(2.0/3.0);
                obj.display_data(source, dataset_train, dataset_test);
                
                fprintf('\nModel ---------------------------------\n');
                layers = [dataset_train.num_features 35 15 10 1];
                network = RegressNet(layers);
                y_min = source.unnormalize_target(0.0);
                y_max = source.unnormalize_target(1.0);
                network.epsilon = 10000 / (y_max - y_min);
                obj.display_model(network);

                %{
                layers = [dataset_train.num_features 35 15 10 3];
                network = ClassNet(layers);
                classes = dataset_train.create_classes(3);
                dataset_train.encode_targets(classes);
                dataset_test.encode_targets(classes);
                obj.display_model(network);
                %}

                fprintf('\nTraining Model.\n');
                num_iters = 500;
                batch_size = 10;
                results = network.train(dataset_train, dataset_test, SGD(0.1, 0.9, 0.5), num_iters, batch_size, @obj.display_training);
                %results = network.train(dataset_train, dataset_test, AdaGrad(0.1, 0.5), num_iters, batch_size, @obj.display_training);
                %results = network.train(dataset_train, dataset_test, AdaDelta(0.5), num_iters, batch_size, @obj.display_training);
                obj.plot(source, network, results);

                fprintf('\nEvaluating model.\n');
                [loss, acc] = network.evaluate(dataset_test);
                obj.display_evaluation(loss, acc);
            end
        end
        function display_data(obj, source, data_train, data_test)
            fprintf('Data Source: %s\n', source.name);
            fprintf('Total features: %d\n', data_train.num_features);
            fprintf('Total entries: %d\n', data_train.num_entries + data_test.num_entries);
            fprintf('Training entries: %d\n', data_train.num_entries);
            fprintf('Test entries: %d\n', data_test.num_entries);
        end
        function display_model(obj, network)
            fprintf('Type: Feedforward Neural Network\n');
            fprintf('Objective: %s\n', network.name);
            fprintf('Layers:\n');
            n_layers = length(network.layers);
            for i = 1:n_layers
                fprintf('\t%d: %d units\n', i, network.layers(i))
            end
        end
        function display_training(obj, results)
            results = cell2mat(cellfun(@(x) cell2mat(x), results, 'un', 0));
            fprintf('[%4d] training [loss=%8.6f acc=%4.2f] validating [loss=%8.6f acc=%4.2f]\n', results(end,:));
            
            gcf;
            subplot(2, 1, 1);
            plot(results(:,1), results(:,3), 'r', results(:,1), results(:,5), 'g');
            ylabel('Accuracy');
            legend('Training Data', 'Test Data', 'Location', 'NorthEastOutside');
            grid('on');
            subplot(2, 1, 2);
            plot(results(:,1), results(:,2), 'r', results(:,1), results(:,4), 'g');
            legend('Training Data', 'Test Data', 'Location', 'NorthEastOutside');
            xlabel('Iteration');
            ylabel('Loss');
            grid('on');
            drawnow();
        end
        function display_evaluation(obj, loss, acc)
            fprintf('Results: [loss=%8.6f acc=%4.2f]\n', loss, acc*100.0);
        end
        function plot(obj, source, network, results)
            results = cell2mat(cellfun(@(x) cell2mat(x), results, 'un', 0));
            
            clf();
            figure(1);
            plot(results(:,1), results(:,3), 'r', results(:,1), results(:,5), 'g');
            title(source.name);
            xlabel('Iteration');
            ylabel('Accuracy');
            legend('Training Data', 'Test Data', 'Location', 'southeast');
            grid('on');
            file_path = sprintf('fig_%s_%s_acc.jpg', lower(source.name), lower(network.name));
            saveas(gcf, file_path);
            close(gcf)

            figure(2);
            plot(results(:,1), results(:,2), 'r', results(:,1), results(:,4), 'g');
            title(source.name);
            xlabel('Iteration');
            ylabel('Loss');
            legend('Training Data', 'Test Data', 'Location', 'northeast');
            grid('on');
            file_path = sprintf('fig_%s_%s_loss.jpg', lower(source.name), lower(network.name));
            saveas(gcf, file_path);
            close(gcf)
            
            figure(3);
            plot(results(:,1), results(:,2), 'r', results(:,1), results(:,4), 'g');
            title(source.name);
            xlabel('Iteration');
            ylabel('Loss');
            legend('Training Data', 'Test Data', 'Location', 'northeast');
            set(gca, 'YScale', 'log');
            grid('on');
            file_path = sprintf('fig_%s_%s_loss_log.jpg', lower(source.name), lower(network.name));
            saveas(gcf, file_path);
            close(gcf)
        end
    end
end

File: Code/Development/NeuralNet/src/python/Experiment.py
# -*- coding: utf-8 -*-
"""
Created on Tue Apr 04 23:10:38 2017

@author: mick
"""

import csv
import matplotlib.pyplot as plt
import numpy as np
import sys
from HousingData import HousingData
from NeuralNetwork.Activations import Activations
from NeuralNetwork.Dataset import Dataset
from NeuralNetwork.Errors import Errors
from NeuralNetwork.Feedforward import FNN
from NeuralNetwork.Optimizers import SGD
from NeuralNetwork.Optimizers import Adadelta
from NeuralNetwork.Optimizers import Adagrad

class Experiment(object):
    def __init__(self):
        """Creates an application for managing overall execution.
        """
        print('')
        print('Loading data.')
        self.__sources = []
        self.__sources.append(HousingData('../../data/art_processed.csv', name='ART'))
        self.__sources.append(HousingData('../../data/redfin_processed.csv', name='GrandRapids'))
        self.__sources.append(HousingData('../../data/kingcounty_processed.csv', name='KingCounty'))
        self.__sources.append(HousingData('../../data/nashville_processed.csv', name='Nashville'))

        self.__figure = plt.figure(0)
        
    def run(self, num_iters):
        """Executes the application.
        """
        for source in self.__sources:
            self.eval_regress_model(source, [35, 15, 10], Adagrad(learning_rate=0.01), num_iters, label='35_15_10')
            self.eval_regress_model(source, [32, 16, 8], Adagrad(learning_rate=0.01), num_iters, label='35_16_8')
        print('Done.')
    
    def eval_class_model(self, source, hidden_layers, optimizer, num_iters, label='classification'):
        print('')
        print('Data  ' + '-'*60)
        data, targets = source.data
        dataset_train, dataset_test = Dataset(data, targets).split(2.0/3.0)
        self.display_data(source, dataset_train, dataset_test)
         
        print('')
        print('Model ' + '-'*60)
        num_classes = 3
        classes = dataset_train.create_classes(num_classes)
        dataset_train.encode_targets(classes)
        dataset_test.encode_targets(classes)
        layers = [dataset_train.num_features] + hidden_layers + [num_classes]
        network = FNN(layers)
        network.name = label
        network.optimizer = optimizer
        network.activation = Activations.Sigmoid
        network.error = Errors.CategoricalCrossEntropy
        network.match = lambda y, t: np.argmax(y) == np.argmax(t)
        self.display_model(network)

        print('')
        print('Training model.')
        log = network.train(
                        dataset_train,
                        dataset_validate=dataset_test,
                        num_iters=num_iters,
                        batch_size=10,
                        output=self.display_training)
        self.plot(source, network, log)
        self.write_csv(source, network, log)
        
        print('')
        print('Evaluating model.')
        results = network.evaluate(dataset_test)
        self.display_evaluation(results)
    
    def eval_regress_model(self, source, hidden_layers, optimizer, num_iters, label='regression'):
        print('')
        print('Data  ' + '-'*60)
        data, targets = source.data
        dataset_train, dataset_test = Dataset(data, targets).split(2.0/3.0)
        self.display_data(source, dataset_train, dataset_test)
        
        print('')
        print('Model ' + '-'*60)
        layers = [dataset_train.num_features] + hidden_layers + [1]
        network = FNN(layers)
        network.name = label
        network.optimizer = optimizer
        network.activation = Activations.Sigmoid
        network.error = Errors.MeanSquared
        network.match = lambda y, t: np.abs(y - t) <= source.normalize_target(10000)
        self.display_model(network)

        print('')
        print('Training model.')
        log = network.train(
                        dataset_train,
                        dataset_validate=dataset_test,
                        num_iters=num_iters,
                        batch_size=10,
                        output=self.display_training)
        self.plot(source, network, log)
        self.write_csv(source, network, log)
        
        print('')
        print('Evaluating model.')
        results = network.evaluate(dataset_test)
        self.display_evaluation(results)
    
    def display_data(self, source, dataset_train, dataset_test):
        print('Data Source: {}'.format(source.name))
        print('Total features: {}'.format(dataset_train.num_features))
        print('Total entries: {}'.format(dataset_train.num_entries + dataset_test.num_entries))
        print('Training entries: {}'.format(dataset_train.num_entries))
        print('Test entries: {}'.format(dataset_test.num_entries))

    def display_model(self, network):
        print('Type: {}'.format(network.name))
        print('Layers:')
        for layer in network.layers:
            print('\t{}'.format(layer))        
        
    def display_training(self, log):
        iters = log['iters']
        train_losses = log['train_losses']
        train_accs = log['train_accs']
        val_losses = log['val_losses']
        val_accs = log['val_accs']
        
        # Print progress to console.
        print_out = '[{: 4d}] '.format(iters[-1])
        print_out += 'training [loss={:09.6f} acc={:05.2f}] '.format(train_losses[-1], train_accs[-1] * 100.0)
        print_out += 'validating [loss={:09.6f} acc={:05.2f}]'.format(val_losses[-1], val_accs[-1] * 100.0)
        print(print_out)

        # Plot progress.
        if not self.__figure is None:
            self.__figure.clf()
            plt.subplot(2, 1, 1)
            plt.ylabel('Accuracy')
            plt.plot(iters, train_accs, 'r', label='Training Data')
            plt.plot(iters, val_accs, 'g', label='Test Data')
            plt.legend(loc=4)
            plt.subplot(2, 1, 2)
            plt.xlabel('Iteration')
            plt.ylabel('Loss')
            plt.yscale('log')
            plt.plot(iters, train_losses, 'r', label='Training Data')
            plt.plot(iters, val_losses, 'g', label='Test Data')
            plt.legend(loc=1)
            plt.pause(0.001)
            plt.draw()

    def display_evaluation(self, results):
        loss = results['loss']
        acc = results['acc']
        print('Results: [loss={:09.6f} acc={:05.2f}]'.format(loss, acc * 100.0))

    def write_csv(self, source, network, log):
        filepath = 'results_{}_{}.csv'.format(source.name.lower(), network.name.lower())
        with open(filepath, 'wb') as output_file:
            writer = csv.writer(output_file)
            writer.writerow(['iter', 'loss', 'acc', 'val_loss', 'val_acc'])
            iters = log['iters']
            train_losses = log['train_losses']
            train_accs = log['train_accs']
            val_losses = log['val_losses']
            val_accs = log['val_accs']
            rows = [list(r) for r in zip(iters, train_losses, train_accs, val_losses, val_accs)]
            for row in rows:
                writer.writerow(row)

    def plot(self, source, network, log):
        """Plots the given results.
        Arguments
            results - A set of results for the execution of the neural network.
        """
        iters = log['iters']
        train_losses = log['train_losses']
        train_accs = log['train_accs']
        val_losses = log['val_losses']
        val_accs = log['val_accs']
        
        plt.figure(1)
        plt.title(source.name)
        plt.plot(iters, train_accs, 'r', label='Training Data')
        plt.plot(iters, val_accs, 'g', label='Test Data')
        plt.xlabel('Iteration')
        plt.ylabel('Accuracy')
        plt.legend(loc=4)
        filepath = 'fig_{}_{}_acc.jpg'.format(source.name.lower(), network.name.lower())
        plt.savefig(filepath)
        plt.close()
        
        plt.figure(2)
        plt.title(source.name)
        plt.plot(iters, train_losses, 'r', label='Training Data')
        plt.plot(iters, val_losses, 'g', label='Test Data')
        plt.xlabel('Iteration')
        plt.ylabel('Loss')
        plt.legend(loc=1)
        filepath = 'fig_{}_{}_loss.jpg'.format(source.name.lower(), network.name.lower())
        plt.savefig(filepath)
        plt.close()
        
        plt.figure(3)
        plt.yscale('log')
        plt.title(source.name)
        plt.plot(iters, train_losses, 'r', label='Training Data')
        plt.plot(iters, val_losses, 'g', label='Test Data')
        plt.xlabel('Iteration')
        plt.ylabel('Loss')
        plt.legend(loc=1)
        filepath = 'fig_{}_{}_loss_log.jpg'.format(source.name.lower(), network.name.lower())
        plt.savefig(filepath)
        plt.close()
